### PR TITLE
feat(workflow): Update Unignored Tooltip

### PR DESCRIPTION
**Pull request by iProgramStuff**
I think this is javascriptz.

Added some casing here to deal with the other conditions that could happen and comments around it that hopefully make it clear (I can remove the comments before merging).

I'm not sure if we have an existing function to format `until` however we'd like to display it to be consistent with other components. Would love to slot that in if so. Otherwise, the format is `2021-01-20T03:59:03+00:00` and I need to know how we want this displayed to create a formatter. 

I made `reason_details: InboxReasonDetails;` required because I was running into issues with destructuring a potentially undefined value in the  constructor of `getToolTipDescription`. There is probably a better way to do this.

### CODE REVIEW

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx [hunk 1]
```diff
@@ -30,6 +30,42 @@ function InboxReason({inbox, fontSize = 'sm'}: Props) {
...
+      // If `user_window` is defined, than `user_count` users affected in `user_window` minutes.
+      // else `user_count` events occured since it was ignored.
+      return t('Was ignored until it affected %(user_count)s user(s)%(user_window)s', {
+        user_count: getCountText(user_count || 0),
+        user_window: user_window
+          ? ' in ' + getDuration((user_window || 0) * 60, 0, true)
+          : '',
+      });
+    } else {
+      return ''; // Shouldn't happen, but wasn't sure of a safe fallback.
```
Comment (taylangocmen) [inline src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:65]: We have a guard for `tooltipDescription` in the jsx so I think this should return `null` here, `''` is also falsy but being explicit here would be better

**Review by taylangocmen — APPROVED** [review]
Other than the fall-through case lgtm

**Comment by priscilawebdev** [comment]
Please is this still WIP?

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx (hunk 1 above)
Comment (iProgramStuff) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:65]: Alright I'll change it to that.

**Comment by iProgramStuff** [comment]
> Please is this still WIP?

Yeah, trying to get some feedback on this as I work on these frontend bits so little.

Still concerned about the formatting of `2021-01-20T03:59:03+00:00` in the first branch of the if statement. Do we have a standard formatter for that?

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx (hunk 1 above)
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:65]:
```suggestion
    if (until) {
      // Was ignored until `until` has passed.
      //`until` format "2021-01-20T03:59:03+00:00"
      return t('Was ignored until %(window)s', {
        until,
      });
    } 
    
    if (count) {
      // Was ignored until `count` events occured
      // If `window` is defined, than `count` events occurred in `window` minutes.
      // else `count` events occured since it was ignored.
      return t('Was ignored until it occured %(count)s time(s)%(window)s', {
        count: getCountText(count || 0),
        window: window ? ' in ' + getDuration((window || 0) * 60, 0, true) : '',
      });
    }
    
      if (user_count) {
      // Was ignored until `user_count` users were affected
      // If `user_window` is defined, than `user_count` users affected in `user_window` minutes.
      // else `user_count` events occured since it was ignored.
      return t('Was ignored until it affected %(user_count)s user(s)%(user_window)s', {
        user_count: getCountText(user_count || 0),
        user_window: user_window
          ? ' in ' + getDuration((user_window || 0) * 60, 0, true)
          : '',
      });
    } 
    
   return ''; // Shouldn't happen, but wasn't sure of a safe fallback.
```

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx [hunk 2]
```diff
@@ -30,6 +30,42 @@ function InboxReason({inbox, fontSize = 'sm'}: Props) {
...
+      return t('Was ignored until %(window)s', {
+        until,
+      });
+    } else if (count) {
+      // Was ignored until `count` events occured
+      // If `window` is defined, than `count` events occurred in `window` minutes.
+      // else `count` events occured since it was ignored.
+      return t('Was ignored until it occured %(count)s time(s)%(window)s', {
+        count: getCountText(count || 0),
+        window: window ? ' in ' + getDuration((window || 0) * 60, 0, true) : '',
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:52]: should we translate 'in' too?

**Comment by priscilawebdev** [comment]
@iProgramStuff I usually use the component DateTime to format date and time, but in this case, I think you should check this file https://github.com/getsentry/sentry/blob/master/src/sentry/static/sentry/app/utils/dates.tsx

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx (hunk 2 above)
Comment (iProgramStuff) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:52]: I'm not sure how the translator works. I was hoping `in` _would_ get translated like this. Otherwise it gets a little more cumbersome because `in` is only present in the string if window exists.

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx (hunk 1 above)
Comment (iProgramStuff) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx]: Thanks! Committed this suggestion and I'll change the final return to account for Taylan's suggestion.

Comment (iProgramStuff) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx]: I actually made this `undefined` because I got a type warning about `tooltipDescription` accepting `string | undefined`

**Comment by iProgramStuff** [comment]
> @iProgramStuff I usually use the component DateTime to format date and time, but in this case, I think you should check this file https://github.com/getsentry/sentry/blob/master/src/sentry/static/sentry/app/utils/dates.tsx

Thanks, I looked at this and ended up using `getFormattedDate(until, 'YYYY/MM/DD HH:mm:ss', { local: true })`.

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx [hunk 3]
```diff
@@ -30,6 +31,46 @@ function InboxReason({inbox, fontSize = 'sm'}: Props) {
...
+    count,
+    window,
+    user_count,
+    user_window,
+  }: InboxReasonDetails) => {
+    if (until) {
+      // Was ignored until `until` has passed.
+      //`until` format: "2021-01-20T03:59:03+00:00"
+      return t('Was ignored until %(window)s', {
+        until: getFormattedDate(until, 'YYYY/MM/DD HH:mm:ss', {local: true}),
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:45]: I noticed that we usually use the format below:

```suggestion
        until: getFormattedDate(until, 'MMM D, YYYY LT', {local: true}),
```

Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:45]: the `local` we should set according to the selected utc.

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx [hunk 4]
```diff
@@ -30,6 +31,46 @@ function InboxReason({inbox, fontSize = 'sm'}: Props) {
       ? `More than ${Math.round(count / EVENT_ROUND_LIMIT)}k`
       : `${count}`;
 
+  const getTooltipDescription = ({
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:34]: we could move this function out of InboxReason or we could also write it like below:


```
function InboxReason({inbox, selection, fontSize = 'sm'}: Props) {
  const {reason, reason_details, date_added: dateAdded} = inbox;

  const getCountText = (count: number) =>
    count > EVENT_ROUND_LIMIT
      ? `More than ${Math.round(count / EVENT_ROUND_LIMIT)}k`
      : `${count}`;

  function getTooltipDescription() {
    const {until, count, window, user_count, user_window} = reason_details;
    const {datetime} = selection;
    const {utc} = datetime;

    if (until) {
      // Was ignored until `until` has passed.
      //`until` format: "2021-01-20T03:59:03+00:00"
      return t('Was ignored until %(window)s', {
        until: getFormattedDate(until, 'MMM D, YYYY LT', {local: !utc}),
      });
    }

    if (count) {
      // Was ignored until `count` events occured
      // If `window` is defined, than `count` events occurred in `window` minutes.
      // else `count` events occured since it was ignored.
      if (window) {
        return tct('Was ignored until it occured ([count]) time(s) in [duration]', {
          count: getCountText(count || 0),
          duration: getDuration((window || 0) * 60, 0, true),
        });
      }

      return tct('Was ignored until it occured ([count]) time(s)', {
        count: getCountText(count || 0),
      });
    }

    if (user_count) {
      // Was ignored until `user_count` users were affected
      // If `user_window` is defined, than `user_count` users affected in `user_window` minutes.
      // else `user_count` events occured since it was ignored.
      return t('Was ignored until it affected %(user_count)s user(s)%(user_window)s', {
        user_count: getCountText(user_count || 0),
        user_window: user_window
          ? ' in ' + getDuration((user_window || 0) * 60, 0, true)
          : '',
      });
    }

    return undefined;
  }
```

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx (hunk 2 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:55]: we could write as below:
```

   if (count) {
      // Was ignored until `count` events occured
      // If `window` is defined, than `count` events occurred in `window` minutes.
      // else `count` events occurred since it was ignored.
      if (window) {
        return tct('Was ignored until it occured ([count]) time(s) in [duration]', {
          count: getCountText(count || 0),
          duration: getDuration((window || 0) * 60, 0, true),
        });
      }

      return tct('Was ignored until it occured ([count]) time(s)', {
        count: getCountText(count || 0),
      });
    }
```

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx [hunk 5]
```diff
@@ -30,6 +31,46 @@ function InboxReason({inbox, fontSize = 'sm'}: Props) {
...
+      //`until` format: "2021-01-20T03:59:03+00:00"
+      return t('Was ignored until %(window)s', {
+        until: getFormattedDate(until, 'YYYY/MM/DD HH:mm:ss', {local: true}),
+      });
+    }
+
+    if (count) {
+      // Was ignored until `count` events occured
+      // If `window` is defined, than `count` events occurred in `window` minutes.
+      // else `count` events occured since it was ignored.
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:52]:
```suggestion
      // else `count` events occurred since it was ignored.
```

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx [hunk 6]
```diff
@@ -30,6 +31,46 @@ function InboxReason({inbox, fontSize = 'sm'}: Props) {
...
+    if (until) {
+      // Was ignored until `until` has passed.
+      //`until` format: "2021-01-20T03:59:03+00:00"
+      return t('Was ignored until %(window)s', {
+        until: getFormattedDate(until, 'YYYY/MM/DD HH:mm:ss', {local: true}),
+      });
+    }
+
+    if (count) {
+      // Was ignored until `count` events occured
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:50]:
```suggestion
      // Was ignored until `count` events occurred
```

File: src/sentry/static/sentry/app/types/dynamicSampling.tsx [hunk 7]
```diff
@@ -17,6 +17,10 @@ export enum DynamicSamplingConditionOperator {
   OR = 'or',
   AND = 'and',
   NOT = 'not',
+  /**
+   * It uses all other operators
+   */
+  ALL = 'all',
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/types/dynamicSampling.tsx:23]: this should not be part of this PR

Comment (iProgramStuff) [reply src/sentry/static/sentry/app/types/dynamicSampling.tsx:23]: Yeah I don't know why these are showing up in the diffs even after merging with master.

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx (hunk 4 above)
Comment (iProgramStuff) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:34]: Switched it to a function. Not sure what the benefit of that is though, could you tell me?

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx [hunk 8]
```diff
@@ -30,6 +31,46 @@ function InboxReason({inbox, fontSize = 'sm'}: Props) {
       ? `More than ${Math.round(count / EVENT_ROUND_LIMIT)}k`
       : `${count}`;
 
+  function getTooltipDescription({
+    until,
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:35]: you have access to the reason details above, so you can use it directly instead of passing it as a parameter

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx [hunk 9]
```diff
@@ -30,6 +31,46 @@ function InboxReason({inbox, fontSize = 'sm'}: Props) {
...
+    count,
+    window,
+    user_count,
+    user_window,
+  }: InboxReasonDetails) {
+    if (until) {
+      // Was ignored until `until` has passed.
+      //`until` format: "2021-01-20T03:59:03+00:00"
+      return t('Was ignored until %(window)s', {
+        until: getFormattedDate(until, 'MMM D, YYYY LT', {local: true}),
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:45]: we should use the selected utc here to set local to true or false.

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx (hunk 4 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx]: we use Babel to transpile the ES6 features, so if you write for example:

`const test = () => {} `

it will be transpiled to:

`var test = function test() {}`

while it could be:

`function test() {}`

because of that, I am preferring to write functions directly. I only use const when I need to use `this` or when I want something to be immutable.

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx (hunk 8 above)
Comment (iProgramStuff) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:35]: Ah, that seems cleaner! Thanks :)

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx (hunk 2 above)
Comment (iProgramStuff) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:55]: Alright, wrote it like this so that translations work better :)

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx (hunk 3 above)
Comment (iProgramStuff) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx]: I'm using `const utc = this.props.utc;` but didn't need to import or wrap anything using `withGlobalSelection`. Is it getting the data from a parent component? Surprised TS is not yelling at me about this.

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx [hunk 10]
```diff
@@ -30,11 +31,55 @@ function InboxReason({inbox, fontSize = 'sm'}: Props) {
       ? `More than ${Math.round(count / EVENT_ROUND_LIMIT)}k`
       : `${count}`;
 
+  function getTooltipDescription() {
+    const {until, count, window, user_count, user_window} = reason_details;
+    if (until) {
+      // Was ignored until `until` has passed.
+      //`until` format: "2021-01-20T03:59:03+00:00"
+      return tct('Was ignored until [window]', {
+        window: <DateTime date={until} dateOnly />,
```
Comment (taylangocmen) [inline src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:40]: @priscilawebdev I think this is fine for the tooltip, we only need to show the day here and we can treat the date as timezone'd like we do in other places in the issue stream.

Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:40]: yes. I think Datetime is fine here too :)

**Review by priscilawebdev — APPROVED** [review]

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx [hunk 11]
```diff
@@ -30,11 +31,55 @@ function InboxReason({inbox, fontSize = 'sm'}: Props) {
...
+
+    return undefined;
+  }
+
   function getReasonDetails(): {
     tagType: React.ComponentProps<typeof Tag>['type'];
     reasonBadgeText: string;
     tooltipText?: string;
-    tooltipDescription?: string;
+    tooltipDescription?: string | React.ReactNode;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:82]: I would use `React.ReactElement `instead of `React.ReactNode `or only `React.ReactNode`, because it already contains the type `string` 😉

File: src/sentry/static/sentry/app/types/index.tsx [hunk 12]
```diff
@@ -685,16 +685,18 @@ export type EventOrGroupType =
...
   date_added?: string;
   reason?: number;
-  reason_details?: {
-    until?: string;
-    count?: number;
-    window?: number;
-    user_count?: number;
-    user_window?: number;
-  };
+  reason_details: InboxReasonDetails;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/types/index.tsx:699]: should we move this prop to the top of the object and leave the optional props in the end?

Comment (iProgramStuff) [reply src/sentry/static/sentry/app/types/index.tsx:699]: Sure, fine with me!

File: src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx (hunk 11 above)
Comment (iProgramStuff) [reply src/sentry/static/sentry/app/components/group/inboxBadges/inboxReason.tsx:82]: Changed it :-)
